### ISSUE TITLE
[REQUEST] Make it possible to join Syntax elements with different stylings.

### ISSUE THREAD

**Issue by aroberge** [Textualize/rich#1829]
**How would you improve Rich?**

The feature request I would be interested is the ability to join `Syntax` items so that different parts could be highlighted in a different theme. The desire to do so is to reproduce the highlighting recently introduced by IPython for indicating the location of the part of code that causes an exception as shown in a recent tweet; image below.  This obviously cannot work if each `Syntax` item appends a newline character by default as noted in #1827 ... However, this feature request goes well beyond what is required to "fix" #1827.

In pseudo-code, and using a line of code from the image below, I would like to be able to do something like:
```python
code = Syntax('return ', "python") + Syntax('x[0][i]', "Text only", theme="error") + Syntax('[0]')
print(code)
```
 

![FI5_l4MWUAAF6gI](https://user-images.githubusercontent.com/629698/149224792-d6b6aa76-df0e-4dd9-908c-2dc5ad8cec26.jpg)

**What problem does it solve for you?**

Being able to reproduce the new styles of IPython tracebacks (including in Jupyter notebooks) to show the location of an error, instead of using a separate line and using `^` characters.

I already do this when showing tracebacks within IDLE (not supported by Rich).

**Comment by willmcgugan** [comment]
Syntax objects aren't really something that can be composed because you can't always correctly highlight a portion of a file.

For instance, if I highlight the keyword "return" it will be in an entirely different color if preceded by "#".

To capture the requirement of highlighting portions of the text, Syntax would need to grow a method to apply additional styles to specific line / character ranges.

Something like the following:

```python
syntax = Syntax(code, "python")
syntax.stylize("on yellow", (lineno, start), (lineno, end)) 
```

**Comment by willmcgugan** [comment]
It probably isn't all that niche. Sounds like a nice feature.

There's still the problem that you can't highlight pieces of the code independently. If you want to be sure not to break the highlighting, you would have to highlight the whole first, and then slice it.

With regards to highlighting with good contrast. I think the best solution would be to calculate new colors based on theme colors. For instance if the background was blended with 20% of red.

**Comment by aroberge** [comment]
By implementing my own parsing on highlighted lines, and using `Text` instead of `Syntax`, I was able to achieve what I wanted.  The advantage I see with the approach I used is that the highlighted text is always visible since all the tokens are shown with the same colour, contrasted with the chosen background (which will be possible to define based on user's preference).

Proof of concept:

1. Original information using `^` to indicate the position of an error. The entire code blocks are highlighted using `Syntax`.

![image](https://user-images.githubusercontent.com/629698/151728566-c6eb4808-b5e5-4ae5-b87f-70900006a546.png)

2. Replacing `^` by text highlighted. Lines that do not contain errors are highlighted using `Syntax`. Lines containing errors are done with `Text`, combining multiple parts.

![image](https://user-images.githubusercontent.com/629698/151728611-d17ce58a-d296-4f66-af66-d9956b80eaf0.png)

3. Same example but done in a Jupyter notebook with a light background.

![image](https://user-images.githubusercontent.com/629698/151728663-28bebd83-c121-4a7a-af04-82f16f70acb3.png)

4. This method also works when multiple subparts need to be highlighted

![image](https://user-images.githubusercontent.com/629698/151729160-6e90f4fa-7839-4099-997b-42175d52b760.png)


For my use case, I **think** that I might stick with my approach, even if the proposed `syntax.stylize()` method is implemented.
